strawberryfields/devicespecs/device_specs.py

```diff
@@ -0,0 +1,144 @@
+    """Abstract base class for backend data"""
+
+    @property
+    @abc.abstractmethod
+    def modes(self) -> int:
+        """The supported number of modes of the device.
+
+        If the device supports arbitrary number of modes, set this to 0.
+
+        Returns:
+            int: number of supported modes
+        """
+
+    @property
+    @abc.abstractmethod
+    def remote(self) -> bool:
+        """Whether the backend supports remote execution.
+
+        Returns:
+            bool: returns ``True`` if the backend supports remote execution.
+        """
+
+    @property
+    @abc.abstractmethod
+    def interactive(self) -> bool:
+        """Whether the backend can be used interactively, that is,
+        the backend state is not reset between engine executions.
+
+        Returns:
+            bool: returns ``True`` if the backend supports interactive use.
+        """
+
+    @property
+    @abc.abstractmethod
+    def primitives(self) -> Set[str]:
+        """The primitive set of quantum operations directly supported
+        by the backend.
+
+        Returns:
+            set[str]: the quantum primitives the backend supports
+        """
+
+    @property
+    @abc.abstractmethod
+    def decompositions(self) -> Dict[str, Dict]:
+        """Quantum operations that are not quantum primitives for the
+        backend, but are supported via specified decompositions.
+
+        This should be of the form
+
+        .. code-block:: python
+
+            {'operation_name': {'option1': val, 'option2': val,...}}
+
+        For each operation specified in the dictionary, the
+        :meth:`~Operation.decompose` method will be called during
+        :class:`Program` compilation, with keyword arguments
+        given by the dictionary value.
+
+        Returns:
+            dict[str, dict]: the quamtum operations that are supported
+            by the backend via decomposition
+        """
+
+    @property
+    def parameter_ranges(self) -> Dict[str, List[List[float]]]:
+        """Allowed parameter ranges for supported quantum operations.
+
+        This property is optional.
+
+        Returns:
+            dict[str, list]: a dictionary mapping an allowed quantum operation
+            to a nested list of the form ``[[p0_min, p0_max], [p1_min, p0_max], ...]``.
+            where ``pi`` corresponds to the ``i``th gate parameter.z
+        """
+        return dict()
+
+    @property
+    def topology(self):
+        """The allowed circuit topology of the backend device as a directed
+        acyclic graph.
+
+        This property is optional; if arbitrary topologies are allowed by the device,
+        this will simply return ``None``.
+
+        Returns:
+            networkx.DAGGraph: a directed acyclic
+        """
+        if self.blackbird_template is None:
+            return None
+
+        # returned DAG has all parameters set to 0
+        bb = blackbird.loads(self.blackbird_template)
+
+        if bb.is_template():
```

Comment:
isn't it confusing that something found under `self.blackbird_template` isn't guaranteed to be a template?

Reply:
Yeah, I don't like the name `blackbird_template` for the property anyway. Some suggestions:

* `DeviceSpec.topology` for the blackbird script

* `DeviceSpec.graph` for the DAG

Reply:
Made the change above :+1: